Bug 1358194 - Add platforms for tests

camd opened the pull request:
This change adds the display of each platform an error test belongs to.  Each platform is the link back to the corresponding job in Treeherder Production.

This required a fair bit of refactoring, so apologies for all the churn there.

I did a few things here:
* Moved the tabulation of counts to a separate event, since it's calculated both from ``configureStore`` and from a separate XHR call.  Trying to prevent  ``fetchTests`` from getting any crazier than it has to be.
* Added ``fetchOptions``.  This gets a mapping of ``optionCollectionHash`` to ``options`` like "debug", "opt", "asan" etc.  Unfortunately, our data model doesn't have a foreign key from the ``job`` ``optionCollectionHash`` to the name of the ``option``.  I may be able to modify the GraphQL Schema to make that join for me at some point, though.  Then it'll just get it with the normal call.
* The ``id`` field returned in the GraphQL response is a hash of the object type plus the number.  So I have to un-hash it to extract the ``job.id``.  I need that to link back to the correct job in Treeherder.  So you'll see some extra steps to make that work in there.
* We now get the push and the jobs at the same time, so no need for separate calls there.  This means I could remove the ``push`` store.
* Needed to add a mapping of platform stored names to the user-read-able text.  I stole this directly from the Treeherder UI so it'll match.

camd commented on src/redux/configureStore.js:
```diff
@@ -108,45 +101,51 @@ async function fetchTests(store, fetchOptions) {
+    .all(errorSummaries.map((log, idx) => getErrorLines(log, urlToJobMap[logUrls[idx]])));
+  // Tests grouped, but could have duplicate test names if the same test failed
+  // in more than one platform.
+  const flatSuites = errorLines.reduce((suiteAcc, result) => Object
```
I tried to combine ``flatSuites`` and ``suites`` below into one ``const`` declaration, but gave up after a while.  I got it close, but since some of the data is really nested, some of the reduces was wiping out objects lower down.  Perhaps I'll get smarter later and be able to figure this out.  :)

camd commented:
@eliperelman : This involved quite a bit of refactoring to get it to work with the expanded data I needed.  Apologies this is so large.  Thanks for all your help!!  :)

eliperelman commented on src/redux/configureStore.js:
```diff
@@ -158,18 +157,69 @@ async function fetchTests(store, fetchOptions) {
+  store.dispatch({
+    type: suitesStore.types.STORE_OPTIONS,
+    payload: {
+      options: options.data.allOptionCollections.reduce((acc, opt) => (
```
I'd keep the paren with the brace here:

```js
.reduce((acc, opt) => ({
  ...acc,
  [opt.optionCollectionHash]: opt.option.name
}), {}),
```

eliperelman commented on src/redux/configureStore.js:
```diff
@@ -158,18 +157,69 @@ async function fetchTests(store, fetchOptions) {
 }
 
 function filterTests(store, { filter, suites, fulltext }) {
+  const myRe = new RegExp(filter, 'i');
```
Can we give this regex a clearer name?

eliperelman commented on src/redux/store.js:
```diff
@@ -1,3 +1,88 @@
 export const { store, actions } = configureStore();
+// Map platform key to a user-friendly platform name
+export const platformMap = {
+  "linux32": "Linux",
```
Replace single quotes with double quotes for consistency.

eliperelman commented on src/ui/Suites.js:
```diff
@@ -90,22 +93,10 @@ export class Suites extends React.Component {
-                    </td>
+                {Object.entries(rows).map(([testName, jobs], rkey) => (
+                  <tr key={rkey}>
+                    <Status/>
```
nit: Insert space between tag name a self closing for consistency.

eliperelman commented on src/ui/Suites.js:
```diff
@@ -90,22 +93,10 @@ export class Suites extends React.Component {
+                {Object.entries(rows).map(([testName, jobs], rkey) => (
+                  <tr key={rkey}>
+                    <Status/>
+                    <Test testName={testName} jobs={jobs} />
```
Since this component is already `Test`, can the property be `name` instead of `testName`?

eliperelman commented on src/ui/Test.js:
```diff
@@ -1,23 +1,51 @@
+          Failed
+        </Badge>
+      </td>
+    )
```
Missing semicolon.

eliperelman commented on src/ui/Test.js:
```diff
@@ -1,23 +1,51 @@
+                <Link
+                  to={`${this.props.treeherder}/#/jobs?repo=mozilla-inbound&revision=${this.props.revision}&selectedJob=${job.jobId}`}
+                  target="_blank">
+                    {job.intermittent && <Icon name="bug" style={{color: '#d78236', verticalAlign: 'text-bottom', marginRight: '.5rem'}}/>}
```
Dedent this line.

eliperelman commented on src/ui/Test.js:
```diff
@@ -1,23 +1,51 @@
+                  to={`${this.props.treeherder}/#/jobs?repo=mozilla-inbound&revision=${this.props.revision}&selectedJob=${job.jobId}`}
+                  target="_blank">
+                    {job.intermittent && <Icon name="bug" style={{color: '#d78236', verticalAlign: 'text-bottom', marginRight: '.5rem'}}/>}
+                    {job.infra && <Icon name="chain-broken" style={{color: '#db3737', verticalAlign: 'text-bottom'}}/>}
```
Dedent this line.

eliperelman commented on src/ui/Test.js:
```diff
@@ -1,23 +1,51 @@
+                  target="_blank">
+                    {job.intermittent && <Icon name="bug" style={{color: '#d78236', verticalAlign: 'text-bottom', marginRight: '.5rem'}}/>}
+                    {job.infra && <Icon name="chain-broken" style={{color: '#db3737', verticalAlign: 'text-bottom'}}/>}
+                  { platformMap[job.buildPlatform.platform] } { this.props.options[job.optionCollectionHash] }
```
Remove extra spaces from inside JSX interpolation braces.

camd commented:
Thanks for the feedback!  I believe I've addressed all the comments.  NOW how much would you pay for this PR?  But wait!  There's a follow-up PR at NO ADDITIONAL COST!

I'm hilarious...
